### PR TITLE
systemd: Support Graceful Reboot for AIO Node

### PR DESCRIPTION
**Pull request by hswong3i**
Ceph AIO installation with single/multiple node is not friendly for loopback mount, especially always get deadlock issue during graceful system reboot. 

We already have `rbdmap.service` with graceful system reboot friendly as below:

    [Unit]
    After=network-online.target
    Before=remote-fs-pre.target
    Wants=network-online.target remote-fs-pre.target
    
    [Service]
    ExecStart=/usr/bin/rbdmap map
    ExecReload=/usr/bin/rbdmap map
    ExecStop=/usr/bin/rbdmap unmap-all
    
This PR introduce:

  - `ceph-mon.target`: Ensure startup after `network-online.target` and before `remote-fs-pre.target`
  - `ceph-*.target`: Ensure startup after `ceph-mon.target` and before `remote-fs-pre.target`
  - `rbdmap.service`: Once all `_netdev` get unmount by `remote-fs.target`, ensure unmap all RBD BEFORE any Ceph components under `ceph.target` get stopped during shutdown
  
The logic is concept proof by <https://github.com/alvistack/ansible-role-ceph_common/tree/develop>; also works as expected with Ceph + Kubernetes deployment by <https://github.com/alvistack/ansible-collection-kubernetes/tree/develop>. No more deadlock happened during graceful system reboot, both AIO single/multiple no de with loopback mount.

Also see: 

  - <https://github.com/ceph/ceph/pull/36776> 
  - <https://github.com/etcd-io/etcd/pull/12259>
  - <https://github.com/cri-o/cri-o/pull/4128>
  - <https://github.com/kubernetes/release/pull/1504>

Fixes: https://tracker.ceph.com/issues/47528
Signed-off-by: Wong Hoi Sing Edison <hswong3i@gmail.com>

image

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by hswong3i** [comment]
@batrick @jtlayton  Latest push should be good enough, well tested with AIO x1 or AIO x3 cluster:

image

**Comment by hswong3i** [comment]
@batrick @jtlayton  Latest push should be good enough, well tested with AIO x1 or AIO x3 cluster:

image

**Comment by jtlayton** [comment]
IIUC, you're doing this so that you can mount cephfs on the host where the daemons may be running (i.e. a loopback mount)? Traditionally, we have suggested that people tag ceph mounts with "_netdev" to ensure that they are unmounted before the network interfaces go down, but I guess that's not sufficient for this use-case.

Is the intention to enable this everywhere, or just in configurations where there are loopback mounts? Do the systemd folks have any clear suggestions for how we should be dealing with this? Having to add dedicated "unmount" services for all different fstypes seems untenable long-term.

**Review by jtlayton — DISMISSED** [review]
The code looks correct for what it does -- just not sure this is the right thing to do.

**Comment by hswong3i** [comment]
> IIUC, you're doing this so that you can mount cephfs on the host where the daemons may be running (i.e. a loopback mount)? Traditionally, we have suggested that people tag ceph mounts with "_netdev" to ensure that they are unmounted before the network interfaces go down, but I guess that's not sufficient for this use-case.
> 
> Is the intention to enable this everywhere, or just in configurations where there are loopback mounts? Do the systemd folks have any clear suggestions for how we should be dealing with this? Having to add dedicated "unmount" services for all different fstypes seems untenable long-term.

Yes, traditionally specify with [`_netdev`](https://www.freedesktop.org/software/systemd/man/systemd.mount.html#_netdev) should be enough for remote mount which ensure unmount happened before `network-online.target` go down. This could be handled automatically by systemd with `remote-fs.target`.

BTW, for loopback mount with [`_netdev`](https://www.freedesktop.org/software/systemd/man/systemd.mount.html#_netdev) we need to secure unmount happened BEFORE BOTH `ceph.target` AND `network-online.target` go down, else we lost the remote service (well, here it become a local service) for making it happen.

With recent push some cleanup as below:
  - `ceph-mon.target`: Ensure startup after `network-online.target`
  - `ceph-*.target`: Ensure startup after `ceph-mon.target`
  - `rbdmap.service`: Once all `_netdev` get unmount by `remote-fs.target`, ensure unmap all RBD BEFORE any Ceph components under `ceph.target` get stopped during shutdown
  - `cephfs-umount.service`: **NEW**. Similar as `rbdmap.service`, once all `_netdev` get unmount by `remote-fs.target`, ensure lazily unmount all CephFS with `umount -lf -a -t ceph` BEFORE any Ceph components under `ceph.target` get stopped during shutdown

Long story short:
  - Keeping `ceph-*.target` startup/shutdown in expected order
  - Unmap RBD and unmount CephFS BEFORE `ceph.target` get stopped

The changes should now be friendly and elegant enough for both traditional remote mount or loopback mount ;-)

image

**Comment by hswong3i** [comment]
Well... `After=/Before=` + `Wants=` combo do the real tricks... Finally could order all `ceph-*.target` startup before `ceph.target`, then move to `rbdmap.service` and `cephfs-umount.service` before `remote-fs.target` correctly:

image

**Comment by hswong3i** [comment]
Wow this even looks much better ;-)

image

**Comment by hswong3i** [comment]
@jtlayton I had been using this PR changes for more then a week with numbers of different cluster setup, and never facing deadlock during graceful system reboot or shutdown. Shall we give a review for here again?

File: systemd/ceph-osd@.service.in [hunk 1]
```diff
@@ -1,8 +1,10 @@
 [Unit]
 Description=Ceph object storage daemon osd.%i
-After=network-online.target local-fs.target time-sync.target ceph-mon.target
-Wants=network-online.target local-fs.target time-sync.target
 PartOf=ceph-osd.target
+After=network-online.target local-fs.target time-sync.target
+Before=ceph-osd.target
```
Comment (batrick) [inline systemd/ceph-osd@.service.in:5]: Does this make sense? Why would this be both `PartOf` and `Before` `ceph-osd.target`?

**Review by batrick — CHANGES_REQUESTED** [review]

File: systemd/ceph-osd@.service.in (hunk 1 above)
Comment (hswong3i) [reply systemd/ceph-osd@.service.in:5]: https://www.freedesktop.org/software/systemd/man/systemd.unit.html#PartOf=
> Configures dependencies similar to Requires=, but limited to stopping and restarting of units. When systemd stops or restarts the units listed here, the action is propagated to this unit. Note that this is a one-way dependency — changes to this unit do not affect the listed units.

https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Before=
> Those two settings configure ordering dependencies between units. If unit foo.service contains the setting Before=bar.service and both units are being started, bar.service's start-up is delayed until foo.service has finished starting up.

Therefore specify both `PartOf` and `Before` could ensure **ALL** `ceph-osd@*.service` are stopped before `ceph-osd.target` get stopped.

As shown in https://github.com/ceph/ceph/pull/36776#issuecomment-684764395, the final result order looks good enough without conflict ;-)

**Review by batrick — CHANGES_REQUESTED** [review]
LGTM. Please update the commit title corresponding to the change  I made to the PR title. See also my usual copy paste:

Please [title your commits according to the conventions of this project](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#commit-title). This is required for your PR to be merged.

See [this article](https://help.github.com/en/articles/changing-a-commit-message) on GitHub on how to amend commits and update your pull request.

**Comment by hswong3i** [comment]
> LGTM. Please update the commit title corresponding to the change I made to the PR title. See also my usual copy paste:
> 
> Please [title your commits according to the conventions of this project](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#commit-title). This is required for your PR to be merged.
> 
> See [this article](https://help.github.com/en/articles/changing-a-commit-message) on GitHub on how to amend commits and update your pull request.

@batrick Updated, please kindly verify ;-)

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
@neha-ojha can you recommend someone from core familiar with systemd to have a glance at these changes?

**Comment by hswong3i** [comment]
> @neha-ojha can you recommend someone from core familiar with systemd to have a glance at these changes?

Maybe @tchaikov @smithfarm @liewegas according to https://github.com/ceph/ceph/commits/master/systemd?

File: systemd/ceph-mon@.service.in [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
 # According to:
 #   http://www.freedesktop.org/wiki/Software/systemd/NetworkTarget
 # these can be removed once ceph-mon will dynamically change network
 # configuration.
 After=network-online.target local-fs.target time-sync.target
-Wants=network-online.target local-fs.target time-sync.target
-PartOf=ceph-mon.target
+Before=ceph-mon.target
+Wants=network-online.target local-fs.target time-sync.target ceph-mon.target
+RequiresMountsFor=/var/lib/ceph
```
Comment (ddiss) [inline systemd/ceph-mon@.service.in:11]: This path is dependent on how the Ceph `mon data` path is configured. Can't we just rely on the `local-fs.target` to mount everything beforehand here?

File: systemd/cephfs-umount.service.in [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+Wants=network-online.target remote-fs-pre.target ceph.target
+
+[Service]
+EnvironmentFile=-@SYSTEMD_ENV_FILE@
+Type=oneshot
+RemainAfterExit=yes
+ExecStop=/usr/bin/umount -lf -a -t ceph
+
+[Install]
+WantedBy=multi-user.target
```
Comment (ddiss) [inline systemd/cephfs-umount.service.in:14]: This doesn't seem to follow regular systemd mount unit configuration.
Why can't we rely on regular systemd fstab handling (with `_netdev` to flag the network dependency)? Is it just the `ceph.target` ordering dependency? If so, could this be achieved with `x-systemd.before/after`?

File: systemd/ceph-mon@.service.in (hunk 2 above)
Comment (hswong3i) [reply systemd/ceph-mon@.service.in:11]: Since we have `mon data = /var/lib/ceph` as default for `/etc/ceph/ceph.conf` unless expressly define and override, setting `RequiresMountsFor=/var/lib/ceph` could also be a reasonable default?

Similar idea, for advanced user that may possibly (well but rare) locate `mon data` into a remote mount point (!?), specifying `local-fs.target` only may not be enough?

In case for advanced user that going to override `mon data`, shall we also assume that they have complete idea and could update here accordingly?

File: systemd/cephfs-umount.service.in (hunk 3 above)
Comment (hswong3i) [reply systemd/cephfs-umount.service.in:14]: > This doesn't seem to follow regular systemd mount unit configuration.

Yes, totally agree, but case doesn't always go ideally...

In my case due to the ordering conflict between ceph/etcd/crio/kublet and `remote-fs-pre.target`/`remote-fs.target`, my libceph always complain about missing ceph mon during unmount. 

Therefore I need a entrypoint to secure all CephFS unmount cleanly **BEFORE** `ceph.target` goes down.

With this PR now we have:
- `_netdev` unmount by systemd fstab during `remote-fs.target`
- Between `remote-fs-pre.target` and `ceph.target`, `cephfs-umount.service` act as a fail-safe to lazy unmount any possible remaining CephFS mount point
- Now `ceph.target` could able to shutdown cleanly

> If so, could this be achieved with x-systemd.before/after?

We better not do so, because even `systemctl restart ceph.target` will also trigger the fail-safe lazy unmount of running CephFS; with `cephfs-umount.service` it only activate during system graceful reboot/shutdown.

Comment (hswong3i) [reply systemd/cephfs-umount.service.in:14]: @ddiss I will give a try again with my production cluster, to see if we could skip `cephfs-umount.service` and still without reboot deadlock.

Comment (hswong3i) [reply systemd/cephfs-umount.service.in:14]: @ddiss Ok if ALL `ceph.target` and its child services could stop AFTER `remote-fs.target` and `remote-fs-pre.target`, no graceful shutdown/reboot deadlock happened. I will remove the `cephfs-umount.service` from PR accordingly:

Screenshot from 2020-09-17 12-57-57

File: systemd/ceph-mon@.service.in (hunk 2 above)
Comment (jan--f) [reply systemd/ceph-mon@.service.in:11]: > In case for advanced user that going to override `mon data`, shall we also assume that they have complete idea and could update here accordingly?

I don't think it's practical to let user alter the systemd unit. Maybe its better to pass this via the `EnvironmentFile=` directive? Though I'm not sure this is available on the `[Unit]` section.

Comment (hswong3i) [reply systemd/ceph-mon@.service.in]: Ok removed

**Comment by batrick** [comment]
@ddiss does this address your concerns?

**Review by batrick — CHANGES_REQUESTED** [review]
Please create a [tracker issue](https://tracker.ceph.com/) so that this issue can be tracked for backporting.

Then, [please annotate the commit which fixes/resolves a Ceph tracker issue with](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#fixes-lines):

`Fixes: http://tracker.ceph.com/issues/...`

This is essential when examining the history of the repository (this commit fixes what) and helps merge scripts identify issues that have been resolved by a merge. See [this article](https://help.github.com/en/articles/changing-a-commit-message) on GitHub on how to amend commits and update your pull request.

**Comment by hswong3i** [comment]
> Please create a [tracker issue](https://tracker.ceph.com/) so that this issue can be tracked for backporting.
> 
> Then, [please annotate the commit which fixes/resolves a Ceph tracker issue with](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#fixes-lines):
> 
> `Fixes: http://tracker.ceph.com/issues/...`
> 
> This is essential when examining the history of the repository (this commit fixes what) and helps merge scripts identify issues that have been resolved by a merge. See [this article](https://help.github.com/en/articles/changing-a-commit-message) on GitHub on how to amend commits and update your pull request.

@batrick Account created at https://tracker.ceph.com and now "Your account was created and is now pending administrator approval."

**Comment by tchaikov** [comment]
> @batrick Account created at https://tracker.ceph.com and now "Your account was created and is now pending administrator approval."

activated

**Comment by hswong3i** [comment]
Ticket created: https://tracker.ceph.com/issues/47528
Commit message updated: https://github.com/ceph/ceph/commit/d88c834ea44bd67cfde0bd11ec4ded079b76d11a

**Review by ddiss — APPROVED** [review]
Thanks for reworking+retesting the patch-set.
I'm far from an expert on systemd units, but this latest round looks good to me.

**Comment by hswong3i** [comment]
> Thanks for reworking+retesting the patch-set.
> I'm far from an expert on systemd units, but this latest round looks good to me.

Thanks you for insisting a better solution, else I may just get satisfied on the original ugly implementation ;-)

Similar changes also backport to other else related PR, most of them looks improved, too~~~

**Review by batrick — APPROVED** [review]

**Comment by hswong3i** [comment]
Backport PR created with `git cherry-pick -x`:
- https://github.com/ceph/ceph/pull/37300
- https://github.com/ceph/ceph/pull/37301

**Comment by zhangyu163** [comment]
@hswong3i  Excuse me, where is the deadlock position? I would be very grateful if you could provide the recurrence steps.
